File: backend/src/CurriculosProIA.App/Helpers/PersistedAnalysisMapper.cs
using System.Text.Json;
using CurriculosProIA.Domain.Dtos;

namespace CurriculosProIA.App.Helpers;

/// <summary>
/// Reconstrói análise e texto do currículo a partir do banco (histórico).
/// </summary>
public static class PersistedAnalysisMapper
{
    public static AnalysisInput ToAnalysisInput(AnaliseCurriculoListItemDto dto)
    {
        var input = new AnalysisInput
        {
            PontosFortes = dto.PontosFortes ?? new List<string>(),
            PontosMelhorar = dto.PontosMelhorar ?? new List<string>(),
            Recomendacoes = dto.Recomendacoes ?? new List<string>(),
            Score = dto.ScoreGeral,
            Habilidades = dto.PalavrasChaveSugeridas?.ToList() ?? new List<string>(),
            Experiencia = string.Empty,
            Formacao = string.Empty
        };

        if (dto.ResultadoCompleto is { ValueKind: JsonValueKind.Object } rc)
        {
            if (rc.TryGetProperty("habilidades", out var hab) && hab.ValueKind == JsonValueKind.Array)
            {
                var fromRc = JsonSerializer.Deserialize<List<string>>(hab.GetRawText()) ?? new List<string>();
                if (fromRc.Count > 0)
                {
                    input.Habilidades = fromRc;
                }
            }

            if (rc.TryGetProperty("experiencia", out var exp) && exp.ValueKind == JsonValueKind.String)
            {
                input.Experiencia = exp.GetString() ?? string.Empty;
            }

            if (rc.TryGetProperty("formacao", out var form) && form.ValueKind == JsonValueKind.String)
            {
                input.Formacao = form.GetString() ?? string.Empty;
            }

            if (rc.TryGetProperty("areaAtuacao", out var area) && area.ValueKind == JsonValueKind.String)
            {
                input.AreaAtuacao = area.GetString();
            }
        }

        if (input.Habilidades.Count == 0 && dto.PalavrasChaveSugeridas?.Count > 0)
        {
            input.Habilidades = dto.PalavrasChaveSugeridas.ToList();
        }

        if (string.IsNullOrWhiteSpace(input.Experiencia) && input.PontosFortes.Count > 0)
        {
            input.Experiencia = string.Join(". ", input.PontosFortes.Take(5));
        }

        return input;
    }

    public static string? GetResumeText(AnaliseCurriculoListItemDto dto) =>
        dto.CurriculosImportados?.ConteudoExtraido?.Trim();

    public static string? GetResumeId(AnaliseCurriculoListItemDto dto) =>
        dto.IdCurriculo;

    public static string? GetSiteId(AnaliseCurriculoListItemDto dto) =>
        dto.IdSiteVagas;
}

File: frontend/src/app/utils/persisted-analysis.mapper.ts
/**
 * Mapeia análise persistida (histórico/API) para o formato usado pelos serviços no analyzer.
 */

interface ResultadoCompletoParsed {
  habilidades?: string[];
  experiencia?: string;
  Experiencia?: string;
  formacao?: string;
  Formacao?: string;
  score?: number;
  areaAtuacao?: string;
  area_atuacao?: string;
}

export function mapPersistedAnalysisToResult(analysis: any): {
  originalText: string;
  analysis: {
    pontosFortes: string[];
    pontosMelhorar: string[];
    experiencia: string;
    formacao: string;
    habilidades: string[];
    recomendacoes: string[];
    score: number;
    areaAtuacao?: string;
  };
  resumeId?: string;
  analysisId?: string;
} {
  const rc = parseResultadoCompleto(analysis?.resultado_completo ?? analysis?.resultadoCompleto);

  const habilidadesFromRc =
    rc?.habilidades && Array.isArray(rc.habilidades) && rc.habilidades.length > 0
      ? rc.habilidades
      : null;

  const habilidades =
    habilidadesFromRc ??
    analysis?.palavras_chave_sugeridas ??
    analysis?.palavrasChaveSugeridas ??
    [];

  const pontosFortes = analysis?.pontos_fortes ?? analysis?.pontosFortes ?? [];
  let experiencia = rc?.experiencia ?? rc?.Experiencia ?? '';
  if (!experiencia && pontosFortes.length > 0) {
    experiencia = pontosFortes.slice(0, 5).join('. ');
  }

  const originalText =
    analysis?.curriculos_importados?.conteudo_extraido ??
    analysis?.curriculosImportados?.conteudoExtraido ??
    analysis?.originalText ??
    analysis?.original_text ??
    '';

  return {
    originalText,
    analysis: {
      pontosFortes,
      pontosMelhorar: analysis?.pontos_melhorar ?? analysis?.pontosMelhorar ?? [],
      experiencia,
      formacao: rc?.formacao ?? rc?.Formacao ?? '',
      habilidades: Array.isArray(habilidades) ? habilidades : [],
      recomendacoes: analysis?.recomendacoes ?? [],
      score: analysis?.score_geral ?? analysis?.scoreGeral ?? rc?.score ?? 0,
      areaAtuacao: rc?.areaAtuacao ?? rc?.area_atuacao
    },
    resumeId: analysis?.id_curriculo ?? analysis?.idCurriculo,
    analysisId: analysis?.id
  };
}

function parseResultadoCompleto(value: unknown): ResultadoCompletoParsed | null {
  if (!value) {
    return null;
  }
  if (typeof value === 'string') {
    try {
      return JSON.parse(value) as ResultadoCompletoParsed;
    } catch {
      return null;
    }
  }
  if (typeof value === 'object') {
    return value as ResultadoCompletoParsed;
  }
  return null;
}
